File: Data Structures & Algorithms/top-k-elements-in-list/submission-1.ts
class Solution {
    /**
     * @param {number[]} nums
     * @param {number} k
     * @return {number[]}
     */
    topKFrequent(nums: number[], k: number): number[] {
        //PQ - min.Heap (from small to big)
        //build the map
        const count = {};
        for(const num of nums){
            count[num] = (count[num] || 0 ) + 1
        }

        //create the min.heap 
        const heap = new MinPriorityQueue((x) => x[1]); //order by 2nd elemnt value 
        for(const [num, cnt] of Object.entries(count)){
            heap.enqueue([num,cnt]);
            if(heap.size() > k){
                heap.dequeue();
            }
        }

        //put into results 
        const res = [];
        for(let i=0; i < k; i++){
            const [num,cnt] = heap.dequeue();
            res.push(num)
        }

        return res; 
    }
}

File: Data Structures & Algorithms/three-integer-sum/submission-1.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        # res = set()
        nums.sort()
        count = defaultdict(int)

        for num in nums:
            count[num] += 1;
        
        res = []
        for i in range(len(nums)):
            count[nums[i]] -= 1
            if i and nums[i] == nums[i-1]:
                #skip duplicate
                continue
            
            for j in range(i+1, len(nums)):
                count[nums[j]] -= 1;
                if j - 1 > i and nums[j] == nums[j-1]:
                    #skip 
                    continue
                target = -(nums[i] + nums[j])
                if count[target] > 0:
                    res.append([nums[i],nums[j],target])
            for j in range(i + 1, len(nums)):
                count[nums[j]] +=1
            
        return res
            